File: algorithms/lab09/python/01-royfloyd/01-royfloyd.py
# SPDX-License-Identifier: BSD-3-Clause


class Task:
    def __init__(self) -> None:
        # numarul maxim de noduri
        self.NMAX = 105

        # n = numar de noduri
        self.n = 0

        # w[x][y] = costul muchiei de la x la y: (x, y, w[x][y])
        # (w[x][y] = 0 - muchia lipseste)
        #
        # In aceasta problema, costurile sunt strict pozitive.
        self.w: list[list[int]] = []

    def solve(self) -> None:
        self.read_input()
        self.write_output(self.compute())

    def read_input(self) -> None:
        with open("in", "r") as fin:
            tokens = fin.read().split()
            if not tokens:
                return

            self.n = int(tokens[0])
            self.w = [[0] * (self.n + 1) for _ in range(self.n + 1)]

            idx = 1
            for x in range(1, self.n + 1):
                for y in range(1, self.n + 1):
                    self.w[x][y] = int(tokens[idx])
                    idx += 1

    def compute(self) -> list[list[int]]:
        return self.royfloyd()

    def royfloyd(self) -> list[list[int]]:
        d = [[0] * (self.n + 1) for _ in range(self.n + 1)]

        # Initializez matricea d folosind matricea w.
        for i in range(1, self.n + 1):
            for j in range(1, self.n + 1):
                d[i][j] = self.w[i][j]

        # Aplic Roy-Floyd:
        for k in range(1, self.n + 1):
            # Fixez nodul intermediar, k.
            for i in range(1, self.n + 1):
                for j in range(1, self.n + 1):
                    if i == j:
                        continue

                    # Daca nu exista drum de la i la k sau de la k la j, sarim peste.
                    if d[i][k] == 0 or d[k][j] == 0:
                        continue

                    # Fixez cele doua capete ale drumului pe care le updatez (i, j).
                    # Daca nu am drum sau daca obtin un cost mai mic intre i, j, updatez.
                    if d[i][j] == 0 or d[i][k] + d[k][j] < d[i][j]:
                        d[i][j] = d[i][k] + d[k][j]

        return d

    def write_output(self, d: list[list[int]]) -> None:
        with open("out", "w") as fout:
            for x in range(1, self.n + 1):
                row = " ".join(str(d[x][y]) for y in range(1, self.n + 1))
                fout.write(row + " \n")


def main() -> None:
    task = Task()
    task.solve()


# [ATENTIE] NU modifica functia main!
if __name__ == "__main__":
    main()

File: algorithms/lab09/rust/01-royfloyd/01-royfloyd.rs
// SPDX-License-Identifier: BSD-3-Clause

use std::fs::File;
use std::io::{BufRead, BufReader, Write};

#[derive(Default)]
struct Task {
    // n = numar de noduri
    n: i32,

    // w[x][y] = costul muchiei de la x la y: (x, y, w[x][y])
    // (w[x][y] = 0 - muchia lipseste)
    //
    // In aceasta problema, costurile sunt strict pozitive.
    w: Vec<Vec<i32>>,
}

impl Task {
    const NMAX: usize = 105;

    pub fn solve(&mut self) {
        self.read_input();
        let result = self.compute();
        self.write_output(&result);
    }

    fn read_input(&mut self) {
        let fin = File::open("in").expect("in");
        let reader = BufReader::new(fin);
        let mut all_tokens: Vec<i32> = Vec::new();
        for line in reader.lines() {
            let line = line.unwrap();
            for tok in line.split_whitespace() {
                all_tokens.push(tok.parse().unwrap());
            }
        }
        self.n = all_tokens[0];
        self.w = vec![vec![0; (self.n + 1) as usize]; (self.n + 1) as usize];
        let mut idx = 1;
        for x in 1..=self.n {
            for y in 1..=self.n {
                self.w[x as usize][y as usize] = all_tokens[idx];
                idx += 1;
            }
        }
    }

    fn compute(&self) -> Vec<Vec<i32>> {
        return self.royfloyd();
    }

    fn royfloyd(&self) -> Vec<Vec<i32>> {
        let mut d = vec![vec![0i32; (self.n + 1) as usize]; (self.n + 1) as usize];

        // Initializez matricea d folosind matricea w.
        for i in 1..=self.n {
            for j in 1..=self.n {
                d[i as usize][j as usize] = self.w[i as usize][j as usize];
            }
        }

        // Aplic Roy-Floyd:
        for k in 1..=self.n {
            // Fixez nodul intermediar, k.
            for i in 1..=self.n {
                for j in 1..=self.n {
                    if i == j {
                        continue;
                    }

                    // Daca nu exista drum de la i la k sau de la k la j, sarim peste.
                    if d[i as usize][k as usize] == 0 || d[k as usize][j as usize] == 0 {
                        continue;
                    }

                    // Fixez cele doua capete ale drumului pe care le updatez (i, j).
                    // Daca nu am drum sau daca obtin un cost mai mic intre i, j, updatez.
                    if d[i as usize][j as usize] == 0
                        || d[i as usize][k as usize] + d[k as usize][j as usize] < d[i as usize][j as usize]
                    {
                        d[i as usize][j as usize] =
                            d[i as usize][k as usize] + d[k as usize][j as usize];
                    }
                }
            }
        }

        return d;
    }

    fn write_output(&self, d: &[Vec<i32>]) {
        let mut fout = File::create("out").expect("out");
        for x in 1..=self.n {
            for y in 1..=self.n {
                write!(fout, "{} ", d[x as usize][y as usize]).unwrap();
            }
            writeln!(fout).unwrap();
        }
    }
}

// [ATENTIE] NU modifica functia main!
fn main() {
    let mut task = Box::new(Task::default());
    task.solve();
}
